File: src/main/java/ru/reosfire/wmlib/guis/inventory/components/ListView.java
package ru.reosfire.wmlib.guis.inventory.components;

import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.inventory.ItemStack;
import ru.reosfire.wmlib.guis.inventory.Gui;
import ru.reosfire.wmlib.text.Replacement;
import ru.reosfire.wmlib.yaml.common.gui.ListViewConfig;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

public abstract class ListView<T> extends GuiComponent
{
    private final ListViewConfig Config;
    private final List<Integer> Indexes = new ArrayList<>();
    private final HashMap<Integer, T> IndexesData = new HashMap<>();

    private IDataProvider<T> Data;
    protected int index;

    private Button backwardButton;
    private Button forwardButton;

    private ListView(ListViewConfig config, Gui gui)
    {
        super(gui);
        Config = config;
        int start = Math.min(Config.StartIndex, Config.EndIndex);
        int end = Math.max(Config.StartIndex, Config.EndIndex);
        int startX = start % 9;
        int endX = end % 9;
        for (int i = start; i <= end; i++)
        {
            int x = i % 9;
            if(x >= startX && x <= endX) Indexes.add(i);
        }
    }
    public ListView(ListViewConfig config, Gui gui, IDataProvider<T> dataProvider)
    {
        this(config, gui);
        Data = dataProvider;
    }

    private void OnMenuItemClicked(InventoryClickEvent event)
    {
        if(Arrays.binarySearch(Indexes.toArray(new Integer[0]), event.getSlot()) < 0) return;
        OnElementClick(IndexesData.get(event.getSlot()), event);
    }
    @Override
    public void register()
    {
        backwardButton = new BackwardButton();
        gui.addComponent(backwardButton);
        forwardButton = new ForwardButton();
        gui.addComponent(forwardButton);
        addClickHandler(this::OnMenuItemClicked);
    }
    @Override
    public void unregister()
    {
        super.unregister();
        gui.removeComponent(forwardButton);
        gui.removeComponent(backwardButton);
    }
    protected void OnElementClick(T element, InventoryClickEvent event)
    {

    }

    public int getPage()
    {
        int maxPage = Data.getSize() / Indexes.size() + (Data.getSize() % Indexes.size() == 0 ? 0 : 1);
        maxPage = Math.max(1, maxPage);
        return Math.abs(index % maxPage);
    }

    @Override
    public void renderTo(org.bukkit.inventory.Inventory inventory, Replacement... replacements)
    {
        List<T> dataPage = getDataPage(getPage());
        for (int i = 0; i < Indexes.size(); i++)
        {
            T data = i < dataPage.size() ? dataPage.get(i) : null;

            ItemStack item = data == null ? null : RenderElement(data);
            inventory.setItem(Indexes.get(i), item);

            IndexesData.put(Indexes.get(i), data);
        }
    }
    private List<T> getDataPage(int page)
    {
        int pageSize = Indexes.size();
        int startIndex = Math.min(pageSize * (page), Data.getSize());
        int endIndex = Math.min(pageSize * (page + 1), Data.getSize());
        return Data.getData(startIndex, endIndex);
    }

    protected abstract ItemStack RenderElement(T element);

    public class BackwardButton extends Button
    {
        public BackwardButton()
        {
            super(Config.BackButton, ListView.this.gui);
        }

        @Override
        protected void onClick(InventoryClickEvent event)
        {
            index--;
            ListView.this.ReRender();
        }
    }
    public class ForwardButton extends Button
    {
        public ForwardButton()
        {
            super(Config.ForwardButton, ListView.this.gui);
        }

        @Override
        protected void onClick(InventoryClickEvent event)
        {
            index++;
            ListView.this.ReRender();
        }
    }
}

File: src/main/java/ru/reosfire/wmlib/guis/inventory/components/ListView.kt
package ru.reosfire.wmlib.guis.inventory.components

import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.ItemStack
import ru.reosfire.wmlib.guis.inventory.Gui
import ru.reosfire.wmlib.text.Replacement
import ru.reosfire.wmlib.yaml.common.gui.ListViewConfig
import java.util.*
import kotlin.math.*

abstract class ListView<T> private constructor(private val Config: ListViewConfig, gui: Gui) : GuiComponent(gui) {
    private val indexes: MutableList<Int> = ArrayList()
    private val indexesData = HashMap<Int, T?>()
    private var dataProvider: IDataProvider<T>? = null
    protected var index = 0
    private val backwardButton: Button = BackwardButton()
    private val forwardButton: Button = ForwardButton()

    init {
        val start = min(Config.startIndex, Config.endIndex)
        val end = max(Config.startIndex, Config.endIndex)
        val startX = start % 9
        val endX = end % 9
        for (i in start..end) {
            val x = i % 9
            if (x in startX..endX) indexes.add(i)
        }
    }

    constructor(config: ListViewConfig, gui: Gui, dataProvider: IDataProvider<T>?) : this(config, gui) {
        this.dataProvider = dataProvider
    }

    private fun onMenuItemClicked(event: InventoryClickEvent) {
        if (Arrays.binarySearch(indexes.toTypedArray(), event.slot) < 0) return
        onElementClick(indexesData[event.slot], event)
    }

    override fun register() {
        gui.addComponent(backwardButton)
        gui.addComponent(forwardButton)
        addClickHandler { event: InventoryClickEvent -> onMenuItemClicked(event) }
    }

    override fun unregister() {
        super.unregister()
        gui.removeComponent(forwardButton)
        gui.removeComponent(backwardButton)
    }

    protected fun onElementClick(element: T?, event: InventoryClickEvent?) {}

    val page: Int
        get() {
            var maxPage = dataProvider!!.size / indexes.size + if (dataProvider!!.size % indexes.size == 0) 0 else 1
            maxPage = max(1, maxPage)
            return abs(index % maxPage)
        }

    override fun renderTo(inventory: Inventory, vararg replacements: Replacement) {
        val dataPage = getDataPage(page)
        for (i in indexes.indices) {
            val data = if (i < dataPage!!.size) dataPage[i] else null
            val item = data?.let { renderElement(it) }
            inventory.setItem(indexes[i], item)
            indexesData[indexes[i]] = data
        }
    }

    private fun getDataPage(page: Int): List<T>? {
        val pageSize = indexes.size
        val startIndex = min(pageSize * page, dataProvider!!.size)
        val endIndex = min(pageSize * (page + 1), dataProvider!!.size)
        return dataProvider!!.getData(startIndex, endIndex)
    }

    protected abstract fun renderElement(element: T): ItemStack

    inner class BackwardButton : Button(Config.backButton, gui) {
        override fun onClick(event: InventoryClickEvent) {
            index--
            this@ListView.reRender()
        }
    }

    inner class ForwardButton : Button(Config.forwardButton, gui) {
        override fun onClick(event: InventoryClickEvent) {
            index++
            this@ListView.reRender()
        }
    }
}